# Table of Contents
## 1. Importing libraries
## 2. Importing data
## 3. Data consistency exercises
## 4. Tasks responses

# 1. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing data

In [2]:
# Import orders and products files
path = r'C:\Users\radav\OneDrive\Documents\Career Foundry\Data Analytics\Immersion\Achievement 4 Python Fundamentals\Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 3. Data consistency exercises

In [5]:
# Create a Dataframe
df_test = pd.DataFrame()

In [6]:
# Create a mixed column type
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed-type data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# Convert numeric values in mix column to string
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Find missing values in products file
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Create subset of missing values in 'product_name' column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
# Create dataframe of products file without missing 'product_name' records
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
# Identify duplicates in products file
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
# Create new dataframe that doesn't include duplicates in products file
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
# Export revised products file
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv')) 

# 4. Task responses

In [22]:
# Q2 Run the df.describe function on the df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## The days_since_prior_order column has a smaller count than all of the other fields, suggesting some missing values for this variable. The min and max numbers look reasonable for all fields (e.g. 7 days of week and 24 hours in a day when zero is counted).

In [23]:
# Q3 Check for mixed-type data in the df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## Q4 There is no mixed-type data in the df_ords dataframe

In [24]:
# Q5 Check for missing values in the df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [25]:
# Q5 Create subset of missing values in 'days_since_prior_order' column
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [26]:
df_nan_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [30]:
# Check the value counts for order_number = 1 to confirm that missing values in days_since_prior_order column relate to customers' first orders where there are no prior orders
df_ords['order_number'].value_counts()[1]

206209

## Q5 The total missing values in the days_since_prior_order column relate to customers' first orders where there are no prior orders. This was confirmed by matching the total count of missing values with the total count of order_number with a value of '1' corresponding to customers' first orders. Both had counts of 206,209.

In [31]:
# Q6 Add a 'customer_status' column to dataframe indicating if orders are from new or existing customers
df_ords['customer_status'] = np.where(df_ords['order_number'] == 1, 'New','Existing')

In [32]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order customer_status  
0                  8                     NaN             New  
1                  7                    15.0        Existing  
2                 12                    21.0        Existing  
3                  7                    29.0        Existing  
4                 15                    28.0        Existing

## Q6 I added a string column to the dataframe to indicate whether orders are from new or existing customers so that it makes more sense why the first order from a customer would have a missing value in the 'days_since_prior_order' column. I opted to leave the NaN values in the 'days_since_prior_order' column because it is accurate information in that there are no prior orders for these records.

In [33]:
# Q7 Run a check for duplicate values in the df_ords dataframe
df_dups_ords = df_ords[df_ords.duplicated()]

In [34]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, customer_status]
Index: []

## Q7 No duplicates were found in the df_ords dataframe.

In [35]:
# Q8 Check current shape of df_ords dataframe 
df_ords.shape

(3421083, 8)

In [36]:
# Q8 Create new dataframe that doesn't include duplicates in orders file
df_ords_no_dups = df_ords.drop_duplicates()

In [37]:
# Q8 Confirm that no duplicates were found and removed in the df_ords dataframe
df_ords_no_dups.shape

(3421083, 8)

## Q8 Confirmation that no duplicates were found and removed in the df_ords dataframe

In [38]:
# Q9 Export revised orders file
df_ords_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))